addons/resource_distribution_controller.go

```diff
@@ -0,0 +1,224 @@
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	err = r.SpokeClient.List(ctx, &templateList, client.InNamespace(r.CurrentNamespace), client.MatchingLabels{
+		utils.CreatedByLabelKey: utils.CreatorMulticlusterOrchestrator,
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to distribute resources to StorageConsumers: %w", err)
+	}
+	if len(templateList.Items) == 0 {
+		logger.Info("No resources to distribute to StorageConsumers.")
+		err = removeFinalizerFromObject(ctx, r.SpokeClient, &addonDeletionlock, ResourceDistributionFinalizer)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	var storageClientMapping *corev1.ConfigMap
+	storageClientMapping, err = utils.GetStorageClientMapping(ctx, r.SpokeClient, r.CurrentNamespace)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to distribute resources to StorageConsumers: %w", err)
+	}
+
+	for k := range storageClientMapping.Data {
+		for consumerIndex := range storageConsumerList.Items {
+			if storageConsumerList.Items[consumerIndex].Status.Client.ID == k {
+				addVRCToConsumers = append(addVRCToConsumers, storageConsumerList.Items[consumerIndex].GetName())
+			}
+		}
+	}
+
+	for _, template := range templateList.Items {
+		logger.Info("Distributing template", "Template", template.GetName())
+		if template.DeletionTimestamp.IsZero() {
+			err = addFinalizerToObject(ctx, r.SpokeClient, &template, ResourceDistributionFinalizer)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		for _, foundConsumer := range storageConsumerList.Items {
+			expectedConsumer := foundConsumer.DeepCopy()
+			vrc := ocsv1alpha1.VolumeReplicationClassSpec{Name: template.Name}
+			index := slices.Index(expectedConsumer.Spec.VolumeReplicationClasses, vrc)
+			if template.DeletionTimestamp.IsZero() {
+				if slices.Contains(addVRCToConsumers, foundConsumer.GetName()) {
+					logger.Info("Adding VRC to StorageConsumer", "VRC", vrc, "StorageConsumer", foundConsumer.GetName())
+					if index == -1 {
+						expectedConsumer.Spec.VolumeReplicationClasses = append(expectedConsumer.Spec.VolumeReplicationClasses, vrc)
+					}
+				} else {
+					logger.Info("StorageConsumer does not need this VRC. Removing VRC from StorageConsumer", "VRC", vrc, "StorageConsumer", foundConsumer.GetName())
+					if index != -1 {
+						expectedConsumer.Spec.VolumeReplicationClasses = slices.Delete(expectedConsumer.Spec.VolumeReplicationClasses, index, index+1)
+					}
+				}
+			} else {
+				logger.Info("Template is being deleted. Removing VRC from StorageConsumer", "VRC", vrc, "StorageConsumer", foundConsumer.GetName())
+				if index != -1 {
+					expectedConsumer.Spec.VolumeReplicationClasses = slices.Delete(expectedConsumer.Spec.VolumeReplicationClasses, index, index+1)
+				}
+			}
+			if !reflect.DeepEqual(expectedConsumer.Spec, foundConsumer.Spec) {
```

Comment:
Since MCO is only concerned about the vrc, how about we check the deepEqual for Spec.VolumeReplicationClass?

Reply:
Done.